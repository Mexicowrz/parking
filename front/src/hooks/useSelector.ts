import { useSelector as nativeUseSelector } from 'react-redux';
import { RootState } from '../store/configureStore';
/**
 * Обертка хука для стора приложения
 * @param  {(state:RootState)=>T} selector
 */
export const useSelector = <T = unknown>(selector: (state: RootState) => T) =>
    nativeUseSelector<RootState, T>(selector);
