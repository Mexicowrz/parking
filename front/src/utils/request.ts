export const getApiUrl = (url: string) =>
  `${window.location.protocol}//${window.location.hostname}${process.env
    .REACT_APP_API_URL!}${process.env.REACT_APP_API_VERSION!}/${url}`;

/**
 * доступные методы запросов на сервер
 */
export type RequestMethod = 'GET' | 'POST' | 'PATCH' | 'DELETE';
/**
 * Организация запросов на серве
 * @param  {string} url
 * @param  {boolean} isSecure безопасный запрос (для авторизованного пользователя)
 * @param  {RequestInit} init
 * @returns Promise<T>
 */
export const request = async <T>(
  url: string,
  isSecure?: boolean,
  init?: RequestInit,
): Promise<T> => {
  const headers = new Headers(init?.headers);
  headers.set('Content-Type', 'application/json');
  if (isSecure) {
    const token = localStorage.getItem('token');
    // если запрос безопасный, а пользователь ниразу не авторизовывался, то сразу его выкидывать без запроса
    if (!token) {
      throw {
        messages: 'unauthorized',
        status: 401,
        key: new Date().getTime(),
      };
    }
    headers.set('token', token!);
  }
  init = {
    ...init,
    headers,
  };
  init.headers = headers;
  const response = await fetch(getApiUrl(url), init);
  // обработать ошибки
  if (response.status < 200 || response.status >= 400) {
    const dt = await response.json();
    if (dt.messages) {
      throw { ...dt, status: response.status };
    }
    throw {
      messages: ['unkown error'],
      key: new Date().getTime(),
      status: response.status,
    };
  }
  if (isSecure) {
    // сохранить обновленный токен
    const newToken = response.headers.get('token');
    if (newToken) localStorage.setItem('token', newToken);
  }
  const data = (await response.json()) as T;
  return data;
};
