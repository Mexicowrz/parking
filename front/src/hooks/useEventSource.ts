import { useEffect } from 'react';
import { getApiUrl } from 'utils/request';

export const useEventSource = <T>(
  url: string,
  onData: (data: T) => void,
  isSecure: boolean = true,
) => {
  if (isSecure) {
    const token = localStorage.getItem('token');
    url = `${url}${url.includes('?') ? '&' : '?'}token=${token}`;
  }
  useEffect(() => {
    const es = new EventSource(`${getApiUrl(url)}`);
    es.onmessage = (ev: MessageEvent) => {
      onData(JSON.parse(ev.data) as T);
    };
    return () => {
      es.close();
    };
  }, [url, onData]);
  return [];
};
