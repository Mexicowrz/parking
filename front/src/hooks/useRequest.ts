import { useCallback, useState } from 'react';
import { useDispatch } from 'react-redux';
import { actions } from 'store/app/actions';
import { request, RequestMethod } from '../utils/request';

/**
 * Определяет функцию-запрос на сервер
 */
export type RequestFn<P> = (body?: P) => Promise<void>;

/**
 * Результат хука useRequest
 */
type UseRequestResult<P, R> = [
  /** данный запроса */
  {
    data?: R;
    isLoading?: boolean;
    error?: string;
  },
  /** функция инициализации запроса */
  RequestFn<P>,
  /** функция сброса данных запроса (возврат к дефолтным значениям) */
  () => void,
];

/**
 * Хук выполнения запроса на сервере
 *
 * @param  {boolean=true} isSecure флаг безопасности запроса (для авторизованных пользователей)
 * @returns UseRequestResult
 */
export const useRequest = <P, R>(
  url: string,
  method: RequestMethod = 'GET',
  isSecure: boolean = true,
): UseRequestResult<P, R> => {
  const [data, setData] = useState<R | undefined>(undefined);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | undefined>(undefined);
  const dispatch = useDispatch();

  // определяем функцию запроса на сервер
  const requestData = useCallback(
    async (body?: P) => {
      setIsLoading(true);
      try {
        const init = {
          method,
          body: JSON.stringify(body),
        };
        setData(await request<R>(url, isSecure, init));
      } catch (error) {
        // в случае неавторизованного запроса (например, стух токен), выбрасывать в логин
        if (error.status === 401) {
          dispatch(actions.logout);
        }
        if (error.messages) setError(error.messages.join(', '));
        else setError(error.toString());
      } finally {
        setIsLoading(false);
      }
    },
    [url, method, isSecure, dispatch],
  );
  // функция сброса состояния хука
  const reset = useCallback(() => {
    setData(undefined);
    setIsLoading(false);
    setError(undefined);
  }, []);

  return [{ data, isLoading, error }, requestData, reset];
};
