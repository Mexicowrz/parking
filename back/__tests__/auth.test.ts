import request from 'supertest';
import { connectDb, closeDb, apiUrl, app } from './jest.hooks';

beforeAll(async () => connectDb());

afterAll(async () => closeDb());

describe('Check authorization', () => {
  describe('POST /login', () => {
    it('login require', async () => {
      return request(app.server)
        .post(apiUrl('auth/login'))
        .send({ username: '', password: 'admin' })
        .expect(500)
        .then((response) => {
          expect(response.body?.messages).toEqual(
            expect.arrayContaining(['username should not be empty']),
          );
        });
    });
    it('password require', async () => {
      return request(app.server)
        .post(apiUrl('auth/login'))
        .send({ username: 'admin', password: '' })
        .expect(500)
        .then((response) => {
          expect(response.body?.messages).toEqual(
            expect.arrayContaining(['password should not be empty']),
          );
        });
    });
    it('wrong password', async () => {
      return request(app.server)
        .post(apiUrl('auth/login'))
        .send({ username: 'admin', password: 'wrong' })
        .expect(403)
        .then((response) => {
          expect(response.body?.messages).toEqual(
            expect.arrayContaining(['wrong_password']),
          );
        });
    });
    it('admin success login', async () => {
      return request(app.server)
        .post(apiUrl('auth/login'))
        .send({ username: 'admin', password: 'admin' })
        .expect('Content-Type', /json/)
        .expect(200)
        .then((response) => {
          expect(response.body).toEqual(expect.anything());
          expect(response.body.token).toEqual(expect.anything());
          expect(response.body.user).toMatchSnapshot();
        });
    });
  });

  describe('GET /current', () => {
    it('unauthorized request', async () => {
      return request(app.server)
        .get(apiUrl('auth/current'))
        .expect(401)
        .then((response) => {
          expect(response.body?.messages).toEqual(
            expect.arrayContaining(['unauthorized']),
          );
        });
    });
    it('admin request', async () => {
      const {
        body: { token },
      } = await request(app.server)
        .post(apiUrl('auth/login'))
        .send({ username: 'admin', password: 'admin' });

      return request(app.server)
        .get(apiUrl('/auth/current'))
        .set('token', token)
        .expect(200)
        .then((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });
  });
});
