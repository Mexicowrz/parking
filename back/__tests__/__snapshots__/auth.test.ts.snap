// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check authorization GET /current admin request 1`] = `
Object {
  "description": null,
  "firstname": "А",
  "flat": null,
  "id": 1,
  "lastname": "Администратор",
  "middlename": null,
  "role": 0,
  "username": "admin",
}
`;

exports[`Check authorization POST /login admin success login 1`] = `
Object {
  "description": null,
  "firstname": "А",
  "flat": null,
  "id": 1,
  "lastname": "Администратор",
  "middlename": null,
  "role": 0,
  "username": "admin",
}
`;
